File: js-bnrtc2/packages/buffer/src/buffer.ts
import "@bfchain/bnrtc2-buffer-typings";
import { encodeUTF8ToBinary, decodeBinaryToUTF8 } from "@bfchain/util-encoding-utf8";

function nextpow2(n: number): number {
    if (n >= Math.pow(2, 32)) {
        return n;
    }

    n = n - 1; // 这个操作决定当n刚好是2的次幂值时，取n，如果没有，则取2n
    n |= n >>> 1;
    n |= n >>> 2;
    n |= n >>> 4;
    n |= n >>> 8;
    n |= n >>> 16;
    n += 1;
    return n;
}

enum WriteDirection {
    PUT = -1,
    PUSH = 1,
}

const MAX_DATA_LEN = Math.pow(2, 32)

export class Bnrtc2Buffer implements Bnrtc2.Bnrtc2Buffer {
    private static HEAD_RESERVE_SIZE: number = 128;
    static from(data: Uint8Array): Bnrtc2Buffer {
        return new Bnrtc2Buffer(data);
    }
    static create(size: number, data?: Uint8Array): Bnrtc2Buffer {
        if (size < 0) {
            throw RangeError(`size(${size}) must > 0`);
        }

        if (data && size < data.length) {
            size = data.length;
        }

        const len = nextpow2(size + Bnrtc2Buffer.HEAD_RESERVE_SIZE);
        const u8a = new Uint8Array(len);
        const buf = new Bnrtc2Buffer(u8a, Bnrtc2Buffer.HEAD_RESERVE_SIZE, Bnrtc2Buffer.HEAD_RESERVE_SIZE);
        if (data) {
            buf._buf.set(data, buf._endOffset);
            buf._endOffset += size;
        }
        return buf;
    }

    private _buf: Uint8Array;
    private _dataOffset: number
    private _endOffset: number;
    private _isPulled: boolean = false;
    private _isClone: boolean = false;
    private _hasCloned: boolean = false;

    private constructor(data: Uint8Array, satrt: number = 0, end: number = data.length) {
        this._buf = data;
        this._dataOffset = satrt;
        this._endOffset = end;
    }

    private extendBuffer(size: number, direction: WriteDirection) {
        const oldData = this.data()
        const oldDataLen = this.length
        const len = nextpow2(oldDataLen + size + Bnrtc2Buffer.HEAD_RESERVE_SIZE);
        this._buf = new Uint8Array(len);
        if (direction == WriteDirection.PUSH) {
            this._dataOffset = Bnrtc2Buffer.HEAD_RESERVE_SIZE;
            this._buf.set(oldData, this._dataOffset);
            this._endOffset = Bnrtc2Buffer.HEAD_RESERVE_SIZE + oldDataLen;
        } else {
            this._dataOffset = Bnrtc2Buffer.HEAD_RESERVE_SIZE + size;
            this._buf.set(oldData, this._dataOffset);
            this._endOffset = Bnrtc2Buffer.HEAD_RESERVE_SIZE + size + oldDataLen;
        }
        this._isClone = false;
        this._hasCloned = false;
    }


    private checkSpace(len: number, direction: WriteDirection): Bnrtc2Buffer {
        if (len + this.length >= MAX_DATA_LEN) {
            throw new RangeError(`data length(${len + this.length}} exceed limit(${MAX_DATA_LEN})`);
        }

        if (this._isClone) {
            throw new RangeError(`clone buffer is read only`);
        }


        if (direction == WriteDirection.PUT) {
            if (len > this._dataOffset || (this._hasCloned && this._isPulled)) {
                this.extendBuffer(len, direction);
            }
        } else if (direction == WriteDirection.PUSH) {
            if (this._endOffset + len > this._buf.length) {
                this.extendBuffer(len, direction);
            }
        }
        return this;
    }

    get length() {
        return (this._endOffset  - this._dataOffset);
    }

    get capacity() {
        return this._buf.length;
    }

    get buffer() {
        return this._buf;
    }
    // 往前插入
    private _put(data: Uint8Array): Bnrtc2Buffer {
        const len = data.length;
        const buf = this.checkSpace(len, WriteDirection.PUT);
        buf._dataOffset -= len;
        buf._buf.set(data, this._dataOffset);
        return buf;
    }

    public put(data: Uint8Array): Bnrtc2Buffer {
        return this._put(data).putU32(data.length)
    }
    public putBool(value: boolean): Bnrtc2Buffer {
        return this.putU8(value ? 1 : 0)
    }
    public putU8(value: number): Bnrtc2Buffer {
        value = value & 0XFF
        const buf = this.checkSpace(1, WriteDirection.PUT);
        buf._dataOffset -= 1;
        buf._buf[this._dataOffset] = value;
        return buf;
    }
    public putU16(value: number): Bnrtc2Buffer {
        value = value & 0XFFFF
        const buf = this.checkSpace(2, WriteDirection.PUT);
        buf._dataOffset -= 2;
        buf._buf[this._dataOffset] = value >> 8;
        buf._buf[this._dataOffset+1] = value;
        return buf;
    }
    public putU32(value: number): Bnrtc2Buffer {
        value >>>= 0
        const buf = this.checkSpace(4, WriteDirection.PUT);
        buf._dataOffset -= 4;
        buf._buf[this._dataOffset] = value >> 24;
        buf._buf[this._dataOffset+1] = value >> 16;
        buf._buf[this._dataOffset+2] = value >> 8;
        buf._buf[this._dataOffset+3] = value;
        return buf;
    }
    public putStr(str: string): Bnrtc2Buffer {
        const u8a = encodeUTF8ToBinary(str);
        return this.put(u8a);
    }

    private _push(data: Uint8Array): Bnrtc2Buffer {
        const len = data.length;
        const buf = this.checkSpace(len, WriteDirection.PUSH);
        buf._buf.set(data, this._endOffset);
        buf._endOffset += len;
        return buf;
    }
    public push(data: Uint8Array): Bnrtc2Buffer {
        return this.pushU32(data.length)._push(data);
    }
    public pushBool(value: boolean): Bnrtc2Buffer {
        return this.pushU8(value ? 1 : 0)
    }
    public pushU8(value: number): Bnrtc2Buffer {
        value = value & 0XFF
        const buf = this.checkSpace(1, WriteDirection.PUSH);
        buf._buf[this._endOffset] = value;
        buf._endOffset += 1;
        return buf;
    }
    public pushU16(value: number): Bnrtc2Buffer {
        value = value & 0XFFFF
        const buf = this.checkSpace(2, WriteDirection.PUSH);
        buf._buf[this._endOffset] = value >> 8;
        buf._buf[this._endOffset+1] = value;
        buf._endOffset += 2;

        return buf;
    }
    public pushU32(value: number): Bnrtc2Buffer {
        value >>>= 0
        const buf = this.checkSpace(4, WriteDirection.PUSH);
        buf._buf[this._endOffset] = value >> 24;
        buf._buf[this._endOffset+1] = value >> 16;
        buf._buf[this._endOffset+2] = value >> 8;
        buf._buf[this._endOffset+3] = value;
        buf._endOffset += 4;

        return buf;
    }
    public pushStr(str: string): Bnrtc2Buffer {
        const u8a = encodeUTF8ToBinary(str);
        return this.push(u8a);
    }

    private _pull(size: number): Uint8Array {
        const u8a = this._peek(size)
        this._dataOffset += size;
        this._isPulled = true;
        return u8a;
    }

    public pull(): Uint8Array {
        const len = this.pullU32()
        return this._pull(len)
    }

    public pullBool(): boolean {
        return this.pullU8() != 0
    }

    public pullU8(): number {
        const u8a = this._pull(1);
        return u8a[0];
    }
    public pullU16(): number {
        const u8a = this._pull(2);
        return (u8a[0] << 8) | u8a[1];
    }
    public pullU32(): number {
        const u8a = this._pull(4);
        // 这里如果使用u8a[0] << 24 会变成有符号
        return ((u8a[0] << 24) | (u8a[1] << 16) | (u8a[2] << 8) | u8a[3])>>>0;
    }
    public pullStr(): string {
        const u8a = this.pull()
        return decodeBinaryToUTF8(u8a);
    }

    private _peek(size: number, offset: number = 0): Uint8Array {
        const len = this._endOffset - this._dataOffset - offset;
        if (size > len) {
            throw new RangeError(`size(${size}) > data length(${len})`);
        }

        return this._buf.subarray(this._dataOffset+offset, this._dataOffset+offset+size);
    }
    public peek(): Uint8Array {
        const len = this.peekU32()
        return this._peek(len, 4)
    }
    public peekBool(): boolean {
        return this.peekU8() != 0
    }
    public peekU8(): number {
        const u8a = this._peek(1);
        return u8a[0];
    }
    public peekU16(): number {
        const u8a = this._peek(2);
        return (u8a[0] << 8) | u8a[1];
    }
    public peekU32(): number {
        const u8a = this._peek(4);
        // 这里如果使用u8a[0] << 24 会变成有符号
        return ((u8a[0] << 24) | (u8a[1] << 16) | (u8a[2] << 8) | u8a[3])>>>0;
    }
    public peekStr(): string {
        const u8a = this.peek()
        return decodeBinaryToUTF8(u8a);
    }

    // not copy, just a dataView
    public data(): Uint8Array {
        return this._buf.subarray(this._dataOffset, this._endOffset);
    }

    public clone(): Bnrtc2Buffer {
        const buf = new Bnrtc2Buffer(this._buf, this._dataOffset, this._endOffset);
        buf._isClone = true;
        this._hasCloned = true;
        return buf;
    }

    public copy(): Bnrtc2Buffer {
        return Bnrtc2Buffer.create(this.length, this.data());
    }
}

File: buffer/buffer.go
package buffer

import (
	"encoding/binary"
	"errors"

	"github.com/guabee/bnrtc/util"
)

type WriteDirection int8

const (
	PUT  WriteDirection = -1
	PUSH WriteDirection = 1
)
const HeadReserveSize = 256
const MaxBufferLen = ^uint32(0)
const ShrinkRatio = 0.75

const (
	InvalidU8  = ^uint8(0)
	InvalidU16 = ^uint16(0)
	InvalidU32 = ^uint32(0)
	InvalidU64 = ^uint64(0)
)

func nextpow2(n uint32) uint32 {
	if n >= MaxBufferLen { // MaxUint32
		return n
	}

	n = n - 1 // 这个操作决定当n刚好是2的次幂值时，取n，如果没有，则取2n
	n |= n >> 1
	n |= n >> 2
	n |= n >> 4
	n |= n >> 8
	n |= n >> 16
	n += 1
	return n
}

type BufferData struct {
	buffer  []byte
	readers int32
}

func (d *BufferData) Len() int {
	return len(d.buffer)
}

func (d *BufferData) Cap() int {
	return cap(d.buffer)
}

type Buffer struct {
	peeker     func(size uint32, offset uint32) ([]byte, error)
	data       *BufferData
	dataOffset uint32
	endOffset  uint32
	isFreezed  bool
}

func IsValidU8(v uint8) bool {
	return v != InvalidU8
}

func IsValidU16(v uint16) bool {
	return v != InvalidU16
}

func IsValidU32(v uint32) bool {
	return v != InvalidU32
}

func IsValidU64(v uint64) bool {
	return v != InvalidU64
}

// read only
func FromBytes(data []byte) *Buffer {
	return &Buffer{
		data: &BufferData{
			buffer:  data,
			readers: 1,
		},
		dataOffset: 0,
		endOffset:  uint32(len(data)),
		isFreezed:  true,
	}
}

func DataLenSize(data []byte) uint32 {
	return uint32(len(data) + 4) // 4 for len field
}

func StringLenSize(str string) uint32 {
	data := util.Str2bytes(str)
	return uint32(len(data) + 4) // 4 for len field
}

func NewBuffer(size uint32, data []byte) *Buffer {
	dataLen := uint32(len(data))
	if dataLen > 0 && size < dataLen {
		size = dataLen
	}

	bufLen := nextpow2(size + HeadReserveSize)
	buf := &Buffer{
		data: &BufferData{
			buffer:  make([]byte, bufLen),
			readers: 1,
		},
		dataOffset: HeadReserveSize,
		endOffset:  HeadReserveSize,
		isFreezed:  false,
	}

	if dataLen > 0 {
		copy(buf.data.buffer[buf.endOffset:], data)
		buf.endOffset += dataLen
	}

	return buf
}

func (buf *Buffer) DataView() []byte {
	if buf == nil {
		return nil
	}
	// not copy, just view
	return buf.data.buffer[buf.dataOffset:buf.endOffset]
}

// copy it, release the delay slice buffer
func (buf *Buffer) DataCopy() []byte {
	if buf == nil {
		return nil
	}
	d := make([]byte, buf.Len())
	copy(d, buf.data.buffer[buf.dataOffset:buf.endOffset])
	return d
}

func (buf *Buffer) Data() []byte {
	if buf.inLowUsage() {
		return buf.DataCopy()
	} else {
		return buf.DataView()
	}
}

func (buf *Buffer) Buffer() *Buffer {
	if buf.inLowUsage() {
		return FromBytes(buf.DataCopy())
	}

	return buf
}

// clone buffer is read(pull/peek) only
func (buf *Buffer) Clone() *Buffer {
	if buf == nil {
		return nil
	}

	newBuf := &Buffer{
		data:       buf.data,
		dataOffset: buf.dataOffset,
		endOffset:  buf.endOffset,
		isFreezed:  buf.isFreezed,
	}
	newBuf.data.readers += 1
	return newBuf
}

func (buf *Buffer) Copy() *Buffer {
	return NewBuffer(buf.Len(), buf.DataView())
}

func (buf *Buffer) Len() uint32 {
	if buf == nil {
		return 0
	}
	return buf.endOffset - buf.dataOffset
}

func (buf *Buffer) Cap() uint32 {
	if buf == nil {
		return 0
	}
	return uint32(buf.data.Cap())
}

func (buf *Buffer) inLowUsage() bool {
	return float64(buf.Len()) < float64(buf.Cap())*ShrinkRatio
}

func (buf *Buffer) Shrink() *Buffer {
	if buf == nil {
		return nil
	}
	return FromBytes(buf.DataCopy())
}

func (buf *Buffer) extendBuffer(size uint32, direction WriteDirection) {
	oldData := buf.DataView()
	oldDataLen := buf.Len()
	buf.data.readers -= 1
	newLen := nextpow2(size + oldDataLen + HeadReserveSize)
	buf.data = &BufferData{
		buffer:  make([]byte, newLen),
		readers: 1,
	}
	if direction == PUSH {
		buf.dataOffset = HeadReserveSize
		copy(buf.data.buffer[HeadReserveSize:], oldData)
		buf.endOffset = HeadReserveSize + oldDataLen
	} else {
		buf.dataOffset = HeadReserveSize + size
		copy(buf.data.buffer[buf.dataOffset:], oldData)
		buf.endOffset = HeadReserveSize + size + oldDataLen
	}
	buf.isFreezed = false
}

func (buf *Buffer) checkSpace(size uint32, direction WriteDirection) error {
	if size >= MaxBufferLen-buf.Len() {
		return errors.New("buffer is full")
	}

	if buf.isFreezed || buf.data.readers > 1 {
		buf.extendBuffer(size, direction)
	} else {
		if direction == PUT {
			if size > buf.dataOffset {
				buf.extendBuffer(size, direction)
			}
		} else if direction == PUSH {
			if buf.endOffset+size > uint32(len(buf.data.buffer)) {
				buf.extendBuffer(size, direction)
			}
		}
	}

	return nil
}

func (buf *Buffer) put(data []byte) *Buffer {
	dataLen := uint32(len(data))
	err := buf.checkSpace(dataLen, PUT)
	if err == nil {
		buf.dataOffset -= dataLen
		copy(buf.data.buffer[buf.dataOffset:], data)
	}
	return buf
}

func (buf *Buffer) Put(data []byte) *Buffer {
	buf.put(data)
	buf.PutU32(uint32(len(data)))
	return buf
}

func (buf *Buffer) PutBool(value bool) *Buffer {
	var v uint8 = 0
	if value {
		v = 1
	}
	buf.PutU8(v)
	return buf
}

func (buf *Buffer) PutU8(value uint8) *Buffer {
	data := []byte{value}
	buf.put(data)
	return buf
}

func (buf *Buffer) PutU16(value uint16) *Buffer {
	data := make([]byte, 2)
	binary.BigEndian.PutUint16(data, value)
	buf.put(data)
	return buf
}

func (buf *Buffer) PutU32(value uint32) *Buffer {
	data := make([]byte, 4)
	binary.BigEndian.PutUint32(data, value)
	buf.put(data)
	return buf
}

func (buf *Buffer) PutU64(value uint64) *Buffer {
	data := make([]byte, 8)
	binary.BigEndian.PutUint64(data, value)
	buf.put(data)
	return buf
}

func (buf *Buffer) PutString(str string) *Buffer {
	data := util.Str2bytes(str)
	buf.Put(data)
	return buf
}

func (buf *Buffer) push(data []byte) *Buffer {
	dataLen := uint32(len(data))
	err := buf.checkSpace(dataLen, PUSH)
	if err == nil {
		copy(buf.data.buffer[buf.endOffset:], data)
		buf.endOffset += dataLen
	}
	return buf
}

func (buf *Buffer) Push(data []byte) *Buffer {
	buf.PushU32(uint32(len(data)))
	buf.push(data)
	return buf
}

func (buf *Buffer) PushBool(value bool) *Buffer {
	var v uint8 = 0
	if value {
		v = 1
	}
	buf.PushU8(v)
	return buf
}

func (buf *Buffer) PushU8(value uint8) *Buffer {
	data := []byte{value}
	buf.push(data)
	return buf
}

func (buf *Buffer) PushU16(value uint16) *Buffer {
	data := make([]byte, 2)
	binary.BigEndian.PutUint16(data, value)
	buf.push(data)
	return buf
}

func (buf *Buffer) PushU32(value uint32) *Buffer {
	data := make([]byte, 4)
	binary.BigEndian.PutUint32(data, value)
	buf.push(data)
	return buf
}

func (buf *Buffer) PushU64(value uint64) *Buffer {
	data := make([]byte, 8)
	binary.BigEndian.PutUint64(data, value)
	buf.push(data)
	return buf
}

func (buf *Buffer) PushString(str string) *Buffer {
	data := util.Str2bytes(str)
	buf.Push(data)
	return buf
}

func (buf *Buffer) _peek(size uint32, offset uint32) ([]byte, error) {
	bufLen := buf.endOffset - buf.dataOffset - offset
	if size > bufLen {
		return nil, errors.New("no enough data")
	}

	return buf.data.buffer[buf.dataOffset+offset : buf.dataOffset+offset+size], nil
}

func (buf *Buffer) peek(size uint32, offset uint32) ([]byte, error) {
	if buf.peeker == nil {
		return buf._peek(size, offset)
	} else {
		return buf.peeker(size, offset)
	}
}

func (buf *Buffer) PeekWithError(offset uint32) ([]byte, error) {
	dataLen, err := buf.peekU32(offset)
	if err != nil {
		return nil, err
	}
	if dataLen > buf.Len()-offset-4 {
		return nil, errors.New("no enough data")
	}

	return buf.peek(dataLen, offset+4)
}

func (buf *Buffer) PeekWithSize(size uint32, offset uint32) ([]byte, error) {
	dataLen := buf.PeekU32(offset)
	if dataLen != size {
		return nil, errors.New("size not match")
	}
	return buf.peek(dataLen, offset+4)
}

func (buf *Buffer) Peek(offset uint32) []byte {
	dataLen, err := buf.peekU32(offset)
	if err != nil {
		return nil
	}

	data, _ := buf.peek(dataLen, offset+4)
	return data
}

func (buf *Buffer) PeekBool(offset uint32) bool {
	data, err := buf.peek(1, offset)
	if err != nil {
		return false
	}
	return data[0] != 0
}

func (buf *Buffer) PeekU8(offset uint32) uint8 {
	data, err := buf.peek(1, offset)
	if err != nil {
		return InvalidU8
	}
	return data[0]
}

func (buf *Buffer) PeekU16(offset uint32) uint16 {
	data, err := buf.peek(2, offset)
	if err != nil {
		return InvalidU16
	}
	return binary.BigEndian.Uint16(data)
}

func (buf *Buffer) peekU32NoError(offset uint32) uint32 {
	return binary.BigEndian.Uint32(buf.data.buffer[buf.dataOffset+offset : buf.dataOffset+offset+4])
}

func (buf *Buffer) peekU32(offset uint32) (uint32, error) {
	data, err := buf.peek(4, offset)
	if err != nil {
		return InvalidU32, err
	}
	return binary.BigEndian.Uint32(data), nil
}

func (buf *Buffer) PeekU32(offset uint32) uint32 {
	value, _ := buf.peekU32(offset)
	return value
}

func (buf *Buffer) PeekU64(offset uint32) uint64 {
	data, err := buf.peek(8, offset)
	if err != nil {
		return InvalidU64
	}

	return binary.BigEndian.Uint64(data)
}

func (buf *Buffer) PeekString(offset uint32) string {
	dataLen, err := buf.peekU32(offset)
	if err != nil {
		return ""
	}

	data, err := buf.peek(dataLen, 4+offset)
	if err != nil {
		return ""
	}

	return util.Bytes2str(data)
}

func (buf *Buffer) consume(size uint32) {
	buf.dataOffset += size
}

func (buf *Buffer) pull(size uint32) []byte {
	data := make([]byte, size)
	copy(data, buf.data.buffer[buf.dataOffset:buf.dataOffset+size])
	buf.consume(size)
	return data
}

func (buf *Buffer) PullWithSize(size uint32) ([]byte, error) {
	_, err := buf.PeekWithSize(size, 0)
	if err != nil {
		return nil, err
	}
	buf.pullU32()
	return buf.pull(size), nil
}

func (buf *Buffer) PullWithError() ([]byte, error) {
	_, err := buf.PeekWithError(0)
	if err != nil {
		return nil, err
	}
	dataLen := buf.pullU32()
	return buf.pull(dataLen), nil
}

func (buf *Buffer) Pull() []byte {
	data := buf.Peek(0)
	if data == nil {
		return nil
	}

	dataLen := buf.pullU32()
	return buf.pull(dataLen)
}

func (buf *Buffer) PullBool() bool {
	data, err := buf.peek(1, 0)
	if err != nil {
		return false
	}
	buf.consume(1)
	return data[0] != 0
}

func (buf *Buffer) PullU8() uint8 {
	data, err := buf.peek(1, 0)
	if err != nil {
		return InvalidU8
	}
	buf.consume(1)
	return data[0]
}

func (buf *Buffer) PullU16() uint16 {
	data, err := buf.peek(2, 0)
	if err != nil {
		return InvalidU16
	}
	buf.consume(2)
	return binary.BigEndian.Uint16(data)
}

func (buf *Buffer) pullU32() uint32 {
	value := buf.peekU32NoError(0)
	buf.consume(4)
	return value
}

func (buf *Buffer) PullU32() uint32 {
	data, err := buf.peek(4, 0)
	if err != nil {
		return InvalidU32
	}
	buf.consume(4)
	return binary.BigEndian.Uint32(data)
}

func (buf *Buffer) PullU64() uint64 {
	data, err := buf.peek(8, 0)
	if err != nil {
		return InvalidU64
	}
	buf.consume(8)
	return binary.BigEndian.Uint64(data)
}

func (buf *Buffer) PullString() string {
	data := buf.Pull()
	if data == nil {
		return ""
	}

	return util.Bytes2str(data)
}

func (buf *Buffer) PullStringWithError() (string, error) {
	data, err := buf.PullWithError()
	if err != nil {
		return "", err
	}

	return util.Bytes2str(data), nil
}

func (buf *Buffer) setPeeker(f func(size uint32, offset uint32) ([]byte, error)) {
	buf.peeker = f
}

type BufferWithError struct {
	*Buffer
	Err error
}

func NewBufferWithError(buf *Buffer) *BufferWithError {
	buffer := &BufferWithError{
		Buffer: buf,
		Err:    nil,
	}

	buf.setPeeker(func(size uint32, offset uint32) ([]byte, error) {
		if buffer.Err != nil {
			return nil, buffer.Err
		}
		data, err := buf._peek(size, offset)
		if err != nil {
			buffer.Err = err
		}
		return data, err
	})

	return buffer
}

func BufferWithErrorFromBytes(data []byte) *BufferWithError {
	return NewBufferWithError(FromBytes(data))
}

func (buf *BufferWithError) IsError() bool {
	return buf.Err != nil
}
